subprojects/gradle/critical-path/api/src/main/kotlin/com/avito/android/critical_path/CriticalPathRegistry.kt

```diff
@@ -0,0 +1,32 @@
+package com.avito.android.critical_path
+
+import com.avito.android.critical_path.internal.PathBuildProvider
+import com.avito.android.gradle.metric.GradleCollector
+import com.avito.kotlin.dsl.isRoot
+import org.gradle.api.Project
+
+public object CriticalPathRegistry {
+
+    public fun addListener(project: Project, listener: CriticalPathListener) {
+        check(project.isRoot()) {
+            "Can consume build critical path only in root project but was ${project.path}"
+        }
+        getProvider(project).addPathListener(listener)
+    }
+
+    private fun getProvider(project: Project): PathBuildProvider {
+        return if (project.extensions.extraProperties.has(listenerExtra)) {
+            project.extensions.extraProperties.get(listenerExtra) as PathBuildProvider
+        } else {
+            val buildListener = PathBuildProvider()
+
+            GradleCollector.initialize(project, listOf(buildListener))
+
+            project.extensions.extraProperties.set(listenerExtra, buildListener)
+
+            buildListener
+        }
+    }
```

Comment:
Why did you call it controversial?
Are there any alternatives?

Reply:
Because any optimization is always a trade-off. If performance is more important, this is reasonable. Now, this is about 0.5s, maybe more. I already know a couple of possible optimizations. After that, we can rethink the decision.